File: DSA_Level-2/HashMap And Heaps/Count Of All Subarrays With Zero Sum/CountOfAllSubarraysWithZeroSum_Solution.java
import java.util.*;

public class Main {

	public static int solution(int[] arr) {
		Map<Integer,Integer> map=new HashMap<>();
		int sum=0;
		map.put(sum,1);
		int count=0;

		for(int a : arr){
			sum+=a;
			if(map.containsKey(sum)==true){
				count+=map.get(sum);
			}

			map.put(sum,map.getOrDefault(sum,0)+1);
		}

		return count;
	}
	
	public static void main(String[] args) {
		Scanner scn = new Scanner(System.in);
		int n = scn.nextInt();
		int[] arr = new int[n];
		for (int i = 0; i < n; i++) {
			arr[i] = scn.nextInt();
		}
		System.out.println(solution(arr));

	}

}

File: DSA_Level-2/HashMap And Heaps/Count Of All Subarrays With Zero Sum/CountOfAllSubarraysWithZeroSum_Solution.js

function largestSubarray(arr){
    
    let map=new Map();

    let sum=0;
    let count=0;
    map.set(sum,1);

    for(let i=0;i<arr.length;++i){
        
        sum+=arr[i];
        if(map.has(sum)){
            count+=map.get(sum);
            map.set(sum,map.get(sum)+1);
        }
        else
            map.set(sum,1);

    }

    console.log(count);

}

largestSubarray([15,-2,2,-8,1,7,10,23]);
